### DISCUSSION TITLE
How to use the interactive debugging mode?

### DISCUSSION THREAD

**Discussion by BeFranke** [PKU-Alignment/safety-gymnasium#106] [General]
I did not find much information on the keyboard-enabled interactive debugging. I understood that I have to construct the environment with `Debug` appended to the name, e.g. `env = gym.make('SafetyPointGoal1Debug-v0', render_mode="human")`

But from that point I'm stuck. Which methods actually starts the input loop? Or do I have to implement that myself?

Thank you for your guidance!

**Comment by muchvo** [comment]
The input loop is mainly controlled by [this method](https://github1s.com/PKU-Alignment/safety-gymnasium/blob/HEAD/safety_gymnasium/bases/underlying.py#L322-L323) and [this method](https://github1s.com/PKU-Alignment/safety-gymnasium/blob/HEAD/safety_gymnasium/agents/racecar.py#L56-L57). We've implemented a debug function for simple robots, which essentially maps your keyboard signals to the action space of the robot. For more complex robots, such as `Ant`, you can implement your own `debug` method within the [Ant class](https://github1s.com/PKU-Alignment/safety-gymnasium/blob/HEAD/safety_gymnasium/agents/ant.py) to adapt to your external devices.

Here's how tasks suffixed with `Debug` are registered, as shown in the [code](https://github1s.com/PKU-Alignment/safety-gymnasium/blob/HEAD/safety_gymnasium/__init__.py#L37-L38). See line 38 and line 100.

Currently, you only need to create an environment instance with the `human` render mode and the suffix "Debug." You can control the `Point`, `Car`, and `Racecar` agents using the `JKLI` keys on the keyboard.

We will provide more detailed explanations in the documentation later on. I appreciate your reminder.

**Reply by BeFranke** [reply]
Thank you for your reply. I just tried it out, but the code:

```
import safety_gymnasium as gym
env = gym.make('SafetyPointGoal1Debug-v0', render_mode="human")
```

neither opens a window nor blocks, am I missing something here?

If I use

```
import safety_gymnasium as gym
env = gym.make('SafetyPointGoal1Debug-v0', render_mode="human")
env.reset()
env.render()
```

I get a non-interactive render of the environment.

How can I proceed to make it interactive? I am running on Windows, if that makes a difference.

**Reply by muchvo** [reply]
This is because a proper environment main loop is needed to run the simulation. This is intrinsic to the nature of Interactive Debug, where your keyboard takes over the dynamic inputs for the robot. You can run your code within a similar loop as demonstrated in [the file](https://github.com/PKU-Alignment/safety-gymnasium/blob/main/examples/env.py) provided.

**Reply by BeFranke** [reply]
Ah, I see now, it works. Thanks a lot for all your help!
